libsolidity/analysis/StaticAnalyzer.cpp

```diff
@@ -348,3 +374,100 @@ bool StaticAnalyzer::visit(FunctionCall const& _functionCall)
 	}
 	return true;
 }
+
+void StaticAnalyzer::checkDoubleStorageAssignment(Assignment const& _assignment)
+{
+	size_t storageToStorageCopies = 0;
+	size_t toStorageCopies = 0;
+	size_t storageByteArrayPushes = 0;
+	size_t storageByteAccesses = 0;
+	auto count = [&](TupleExpression const& _lhs, TupleType const& _rhs, auto _recurse) -> void {
+		TupleType const& lhsType = dynamic_cast<TupleType const&>(*type(_lhs));
+		TupleExpression const* lhsResolved = dynamic_cast<TupleExpression const*>(resolveOuterUnaryTuples(&_lhs));
+
+		solAssert(lhsResolved && lhsResolved->components().size() == lhsType.components().size());
+		if (lhsType.components().size() != _rhs.components().size())
+		{
+			solAssert(m_errorReporter.hasErrors(), "");
+			return;
+		}
+
+		for (auto&& [index, componentType]: lhsType.components() | ranges::views::enumerate)
+		{
+			if (ReferenceType const* ref = dynamic_cast<ReferenceType const*>(componentType))
+			{
+				if (ref && ref->dataStoredIn(DataLocation::Storage) && !ref->isPointer())
```

Comment:
```suggestion
				if (ref->dataStoredIn(DataLocation::Storage) && !ref->isPointer())
```
`ref` must exist here as